src/webgpu/api/operation/storage_texture/read_only.spec.ts

```diff
@@ -0,0 +1,378 @@
+
+TODO:
+- Test the use of read-only storage textures in vertex and fragment shaders
+- Test 1D and 3D textures
+- Test bgra8unorm with 'bgra8unorm-storage'
+- Test resource usage transitions with read-only storage textures
+`;
+
+import { makeTestGroup } from '../../../../common/framework/test_group.js';
+import { unreachable, assert } from '../../../../common/util/util.js';
+import { Float16Array } from '../../../../external/petamoriken/float16/float16.js';
+import {
+  ColorTextureFormat,
+  kColorTextureFormats,
+  kTextureFormatInfo,
+} from '../../../format_info.js';
+import { GPUTest } from '../../../gpu_test.js';
+
+function ComponentCount(format: ColorTextureFormat): number {
+  switch (format) {
+    case 'r32float':
+    case 'r32sint':
+    case 'r32uint':
+      return 1;
+    case 'rg32float':
+    case 'rg32sint':
+    case 'rg32uint':
+      return 2;
+    case 'rgba32float':
+    case 'rgba32sint':
+    case 'rgba32uint':
+    case 'rgba8sint':
+    case 'rgba8uint':
+    case 'rgba8snorm':
+    case 'rgba8unorm':
+    case 'rgba16float':
+    case 'rgba16sint':
+    case 'rgba16uint':
+      return 4;
+    default:
+      unreachable();
+      return 0;
+  }
+}
+
+class F extends GPUTest {
+  InitTextureAndGetExpectedOutputBufferData(
+    storageTexture: GPUTexture,
+    format: ColorTextureFormat
+  ): ArrayBuffer {
+    const bytesPerBlock = kTextureFormatInfo[format].bytesPerBlock;
+    assert(bytesPerBlock !== undefined);
+
+    const width = storageTexture.width;
+    const height = storageTexture.height;
+    const depthOrArrayLayers = storageTexture.depthOrArrayLayers;
+
+    const texelData = new ArrayBuffer(bytesPerBlock * width * height * depthOrArrayLayers);
+    const texelTypedData = this.GetTypedArrayBufferForTexelData(texelData, format);
+    const componentCount = ComponentCount(format);
+    const outputBufferData = new ArrayBuffer(4 * 4 * width * height * depthOrArrayLayers);
+    const outputBufferTypedData = this.GetTypedArrayBufferForOutputBufferData(
+      outputBufferData,
+      format
+    );
+
+    const SetData = (
+      texelValue: number,
+      outputValue: number,
+      texelDataIndex: number,
+      component: number
+    ) => {
+      const texelComponentIndex = texelDataIndex * componentCount + component;
+      texelTypedData[texelComponentIndex] = texelValue;
+      const outputTexelComponentIndex = texelDataIndex * 4 + component;
+      outputBufferTypedData[outputTexelComponentIndex] = outputValue;
+    };
+    for (let z = 0; z < depthOrArrayLayers; ++z) {
+      for (let y = 0; y < height; ++y) {
+        for (let x = 0; x < width; ++x) {
+          const texelDataIndex = z * width * height + y * width + x;
+          outputBufferTypedData[4 * texelDataIndex] = 0;
+          outputBufferTypedData[4 * texelDataIndex + 1] = 0;
+          outputBufferTypedData[4 * texelDataIndex + 2] = 0;
+          outputBufferTypedData[4 * texelDataIndex + 3] = 1;
```

Comment:
This is slightly surprising, alpha defaults to 1 if it is not present in the formtat? Ok